collectors/plugins.d/plugins_d.c

```diff
@@ -521,6 +521,70 @@ static void pluginsd_worker_thread_cleanup(void *arg) {
     }
 }
 
+#define SERIAL_FAILURES_THRESHOLD 10
+static void pluginsd_worker_thread_handle_success(struct plugind *cd) {
+    if (likely(cd->successful_collections)) {
+        sleep((unsigned int) cd->update_every);
+        return;
+    }
+
+    if(likely(cd->serial_failures <= SERIAL_FAILURES_THRESHOLD)) {
+        info("'%s' (pid %d) does not generate useful output but it reports success (exits with 0). %s.",
+            cd->fullfilename, cd->pid,
+            cd->enabled ?
+                "Waiting a bit before starting it again." :
+                "Will not start it again - it is now disabled.");
+        sleep((unsigned int) (cd->update_every * 10));
+        return;
+    }
+
+    if (cd->serial_failures > SERIAL_FAILURES_THRESHOLD) {
+        error("'%s' (pid %d) does not generate useful output, although it reports success (exits with 0)."
+            "We have tried to collect something %zu times - unsuccessfully. Disabling it.",
+            cd->fullfilename, cd->pid, cd->serial_failures);
+        cd->enabled = 0;
+        return;
+    }
+
+    return;
+}
+
+static void pluginsd_worker_thread_handle_error(struct plugind *cd, int code) {
+    if (code == -1) {
+        info("'%s' (pid %d) was killed with SIGTERM. Disabling it.", cd->fullfilename, cd->pid);
+        cd->enabled = 0;
+        return;
+    }
+
+    if (!cd->successful_collections) {
+        error("'%s' (pid %d) exited with error code %d and haven't collected any data. Disabling it.",
+            cd->fullfilename, cd->pid, code);
+        cd->enabled = 0;
+        return;
+    }
+
+    if (cd->serial_failures <= SERIAL_FAILURES_THRESHOLD) {
+        error("'%s' (pid %d) exited with error code %d, but has given useful output in the past (%zu times). %s",
+            cd->fullfilename, cd->pid, code, cd->successful_collections,
+            cd->enabled ?
```

Comment:
Hm, i think the whole handle plugin exit/restart handling logic is a little bit lame.
For example why do we need to `sleep((unsigned int) (cd->update_every * 10)` if `!cd->enabled` ?

should we even pass `cd` to handle func if it is disabled after `pluginsd_process`? :thinking: 	

i suggest to dig into it and see if there is any logic problems. Could you do it @emmrk?

Reply:
^^ this should be another PR tho

Reply:
Well yes, there's at least this one unnecessary `sleep`... I've made it into an issue (https://github.com/netdata/netdata/issues/6402) so that it's not lost (is that a good idea?). I'll probably look into it later